### PR TITLE
Register plugins at the right time, once

### PR DESCRIPTION
**Pull request by mklim**
## Description

Currently we're automatically registering plugins both when the
FlutterEngine is constructed and in the `flutter create` template, when
FlutterActivity#configureFlutterEngine is called. The initial
registration is too early to contain a reference to the activity and the
second registration can cause problems in some plugins.

This alters the flow so automatic registration happens in two discrete
places, and contains the `activity` in its first and only call for most
apps.

1. We're no longer automatically registering plugins on `FlutterEngine`
in any of our activities/fragments at construction time. But since the
FlutterEngine default constructor still automatically registers plugins,
anyone constructing the engine themselves (for example, in a service) is
still going to get automatic registration at `FlutterEngine`
instantiation time.
2. We now automatically register plugins in the base `FlutterActivity`'s
`configureFlutterEngine` hook. Anyone using `FlutterActivity` (or
`FlutterFragment`) should be automatically registered once that hook is
called. Right now the `flutter create` template overrides the base class
method with a subclass that registers everything manually in the same
spot. But with this in place we can safely recommend to remove the
subclass and rely on this hook to automatically register going forward.
Registering at this time means `activity` is set correctly.

## Related Issues

- Fixes flutter/flutter#49365 in the common case. flutter/engine#15956 is also needed to prevent duplicate registration in all cases.
- Fixes flutter/flutter#47921
- Fixes FirebaseExtended/flutterfire#1669

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java [hunk 1]
```diff
@@ -223,7 +225,7 @@ void onAttach(@NonNull Context context) {
     // FlutterView.
     Log.v(TAG, "No preferred FlutterEngine was provided. Creating a new FlutterEngine for"
         + " this FlutterFragment.");
-    flutterEngine = new FlutterEngine(host.getContext(), host.getFlutterShellArgs().toArray());
+    flutterEngine = new FlutterEngine(host.getContext(), /*automaticallyRegisterPlugins=*/true);
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java]: Whoops, ate a param here. Patching it now.

Comment (mklim) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java]: Fixed. Also added a file that wasn't included in the last commit.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 2]
```diff
@@ -950,4 +957,28 @@ public void onFlutterUiNoLongerDisplayed() {
     // no-op
   }
 
+  /**
+   * Registers all plugins that an app lists in its pubspec.yaml.
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: minor nit: plugins that support the android platform

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 3]
```diff
@@ -950,4 +957,28 @@ public void onFlutterUiNoLongerDisplayed() {
...
+   * <p>
+   * The Flutter tool generates a class called GeneratedPluginRegistrant, which includes the code
+   * necessary to register every plugin in the pubspec.yaml with a given {@code FlutterEngine}.
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: ditto

File: shell/platform/android/io/flutter/embedding/engine/FlutterEngine.java [hunk 4]
```diff
@@ -152,6 +151,17 @@ public FlutterEngine(@NonNull Context context, @Nullable String[] dartVmArgs) {
...
+   * Same as {@link #FlutterEngine(Context)} with added support for passing Dart
+   * VM arguments and avoiding automatic plugin registration.
+   * <p>
+   * If the Dart VM has already started, the given arguments will have no effect.
+   */
+  public FlutterEngine(@NonNull Context context, @Nullable String[] dartVmArgs, boolean automaticallyRegisterPlugins) {
+    this(context, FlutterLoader.getInstance(), new FlutterJNI(), dartVmArgs, automaticallyRegisterPlugins);
+  }
+
+
```
Comment (blasten) [inline shell/platform/android/io/flutter/embedding/engine/FlutterEngine.java]: minor nit: extra line

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 2 above)
Comment (mklim) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: I think that would be redundant to point out here. This is already a Javadoc comment in the Android embedding specifically and not a generic part of the engine's C++ source code, so I don't think it's worth pointing out that this logic is Android-only. If you feel strongly about it I can update it though.

Comment (blasten) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Not at all. 🙂Although, if someone wants to understand why a plugin isn't added to the `GeneratedPluginRegistrant`, the comment might be useful. 

The current definition of supporting the Android platform is a plugin that contains `android/build.gradle`.

https://api.flutter.dev/javadoc/io/flutter/app/FlutterActivity.html.

**Comment by scognito** [comment]
Is there a workaround to fix this issue on stable channel? I'm using Google maps and Firebase and getting notifications twice

**Comment by mklim** [comment]
@scognito there isn't a good workaround on stable yet, sorry. Because of how this is broken there isn't an obvious workaround that doesn't cause a secondary issue. I'm going to see if this can be hotfixed, but for now it's only rolled into the master channel.

**Comment by scognito** [comment]
Switching to master could be a good temporary solution for me. 
Once switched to master, and after a flutter clean and build, do I need to change something else?
Thanks for your efforts.

**Pull request merged**